sourcerer-core/src/main/java/org/elder/sourcerer2/ExpectedVersion.kt

```diff
@@ -0,0 +1,127 @@
+        private fun mergeOrdered(
+                baseVersion: ExpectedVersion,
+                newVersion: ExpectedVersion
+        ): ExpectedVersion {
+            return when (baseVersion) {
+                ExpectedVersion.Any ->
+                    // Any is always compatible with other expected version
+                    newVersion
+                ExpectedVersion.AnyExisting -> when (newVersion) {
+                    ExpectedVersion.AnyExisting,
+                    ExpectedVersion.Any -> ExpectedVersion.anyExisting()
+                    ExpectedVersion.NotCreated -> throw ConflictingExpectedVersionsException(
+                            "Cannot merge 'any existing' with 'not created'",
+                            baseVersion, newVersion)
+                    is ExpectedVersion.Exactly -> newVersion
+                }
+                is ExpectedVersion.Exactly -> when (newVersion) {
+                    is ExpectedVersion.Exactly -> {
+                        if (baseVersion.streamVersion != newVersion.streamVersion) {
+                            throw ConflictingExpectedVersionsException(
+                                    "Cannot merge 'exactly "
+                                            + baseVersion.streamVersion
+                                            + "' with 'exactly "
+                                            + newVersion.streamVersion
+                                            + "'",
+                                    baseVersion, newVersion)
+                        }
+                        baseVersion
+                    }
+                    ExpectedVersion.NotCreated ->
+                        throw ConflictingExpectedVersionsException(
+                                "Cannot merge 'exactly "
+                                        + baseVersion.streamVersion
+                                        + "' with 'not created'",
+                                baseVersion, newVersion)
+                    else ->
+                        throw RuntimeException(
+                                "Internal error, unexpected version combination merging "
+                                        + baseVersion
+                                        + " with " + newVersion)
+
+                }
+                ExpectedVersion.NotCreated -> when (newVersion) {
+                    ExpectedVersion.NotCreated -> baseVersion
+                    else ->
+                        throw RuntimeException(
+                                "Internal error, unexpected version combination merging "
+                                        + baseVersion
+                                        + " with " + newVersion)
+                }
+            }
+        }
+    }
+}
+
+private fun ExpectedVersion.ordinal(): Int {
+    return when (this) {
+        ExpectedVersion.Any -> 0
+        ExpectedVersion.AnyExisting -> 1
+        is ExpectedVersion.Exactly -> 2
+        ExpectedVersion.NotCreated -> 3
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Report magic numbers. Magic number is a numeric literal that is not defined as a constant and hence it's unclear what the purpose of this number is. It's better to declare such numbers as constants and give them a proper name. By default, -1, 0, 1, and 2 are not considered to be magic numbers. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.elder.org/coding_rules#rule_key=detekt-kotlin%3AMagicNumber)